sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -416,8 +418,8 @@ object ColumnPruning extends Rule[LogicalPlan] {
     case w: Window if w.windowExpressions.isEmpty => w.child
 
     // Eliminate no-op Projects
-    case p @ Project(_, child) if sameOutput(child.output, p.output) => child
-
+    case p @ Project(_, child) if sameOutput(child.output, p.output) =>
+      if (child.isInstanceOf[CatalogRelation]) p else child
```

Comment:
This works in your example case, but will fail for a more complex query.

The problem is that the transform on L199 also modifies parts of the tree, that are totally unrelated to the projection at hand. In this case right side of the join. This can currently only happen in case of a self-join on a CTE, such a tree can contain duplicate attribute ids because we expand the tree after we have analyzed the CTE.